### DISCUSSION TITLE
Enabling Schema Awarenesss breaks simplePagination

### DISCUSSION THREAD

**Discussion by Svarto** [urql-graphql/urql#1041] [Q&A] [answered]
I have a working infinite scrolling list in React Native with urql simplePagination. When I added the schema to the app following the documentation, the simplePagination breaks and it doesn't append new items to the old items. It just returns the new items once, and then stops executing.

The Query:
```
  const [offset, setOffset] = useState(0);

  const [result, reexecuteQuery] = useQuery({
    query: GET_RECIPE_FEED,
    variables: { offset },
  });
```

The onEndReached (bold) is where the offset is updated to fetch more results:
```
      <FlatList
        style={styles.flatContainer}
        data={result.data.recipe.filter((obj) => obj)}
        removeClippedSubviews={true}
        renderItem={renderItem}
        ListFooterComponent={
          result.fetching && result.data !== undefined ? LoadingFooter : Footer
        }
        refreshing={result.fetching}
        onRefresh={() => {
          setOffset(0);
          reexecuteQuery({
            requestPolicy: "cache-and-network",
          });
        }}
        keyExtractor={(item) => item.id.toString()}
        onEndReachedThreshold={0.5}
        **onEndReached={() => {
          // The fetchMore method is used to load new data and add it
          // to the original query we used to populate the list

          setOffset(result.data.recipe.length);
        }}**
      />
```

The client resolvers with simplePagination:
```
export const CacheExchange = cacheExchange({
  resolvers: {
    Query: {
      recipe: simplePagination({ offsetArgument: "offset" }),
    },
  },
```

Is this expected behaviour or can I do something to use both Schema Awareness and simplePagination at once?

**Comment by kitten** [comment, marked as answer]
Is it correct that you added this pagination resolver to `Query.recipe`? That sounds like a singular item and the query says something about a feed, so I think you may have been meaning to add the resolver to some kind of `Recipe` field instead?

**Reply by Svarto** [reply]
@kitten, I believe it is correct, but I am still learning so might be wrong. When using urql-devtools this is what the cache looks like:

Toplevel:
![image](https://user-images.githubusercontent.com/24279217/95191338-27eeaa80-07c0-11eb-9aaa-7e5035b6e72d.png)

When offset is 0, so the "first load of the list": 
![image](https://user-images.githubusercontent.com/24279217/95191411-43f24c00-07c0-11eb-9af1-a37635ff0ddb.png)

And then when I scroll down, I expect the second load to include the first, i.e. append to offset 0 the results from when offset 5 (like it does when schema is not enabled). But as you can see, the first recipe is a different one:
![image](https://user-images.githubusercontent.com/24279217/95191534-6b491900-07c0-11eb-9aca-869696aff6b2.png)

I have a weird naming convention, I should have called my postgres tables "recipes" as it is plural, but I haven't gotten to changing it yet.

This is the query for the feed:
```
// Used in the discovery feed
export const GET_RECIPE_FEED = gql`
  query GetRecipeFeed($offset: Int) {
    recipe(
      order_by: { created_at: desc }
      limit: 5
      offset: $offset
      where: { draft: { _eq: false } }
    ) {
      id
      title
      description
      updated_at
      dishtype
      difficulty
      duration
      recipe_tags {
        tag {
          tag
        }
      }
      images {
        id
        filename
        user_id
        token
        folder
        height
        width
      }
      user {
        id
        username
        avatars {
          id
          filename
          folder
          token
          user_id
        }
      }
    }
  }
`;

```

**Reply by kitten** [reply]
Ok that's certainly weird; If the UI isn't showing you the combined list then it's clearly as if the pagination resolver wasn't doing anything. Without more info I'm not really sure what's going on though. Have you tried wrapping the resolver to see what it's returning?

```js
const recipesResolver = simplePagination({ offsetArgument: "offset" });

export const CacheExchange = cacheExchange({
  resolvers: {
    Query: {
      recipe: (parent, fieldArgs, cache, info) => {
        const result = recipesResolver(parent, fieldArgs, cache, info);
        console.log(result, JSON.stringify(info, null, 2))
        return result;
      }
    },
  },
});
```

**Reply by jjstanton** [reply]
@kitten I'm also having the same behavior when using `simplePagination` and schema awareness. At first I thought it could be the `makeDefaultStorage` I am using but after some troubleshooting, both the pagination and storage were working after removing the schema. 

I have tried wrapping the resolver as you suggested and nothing was logged to the console with the schema being added. As soon as I remove the schema, the logs were pouring in. Seems like the resolver is completely ignored when the schema is in place.

Here is my setup (with the wrapper):

```
const storage = makeDefaultStorage({
    idbName: 'graphcache-v3', // The name of the IndexedDB database
    maxAge: 7, // The maximum age of the persisted data in days
});

const vxFixtureResolver = simplePagination({ offsetArgument: "offset", limitArgument: "limit" })

const cache = cacheExchange({
    // schema: (schema as unknown) as IntrospectionQuery,
    storage,
    keys: {
        vwCompetitions: (data: any) => data.ID,
        vwFixtures: (data: any) => data.FixtureID,
        app_Logs: (data: any) => data.ID
    },
    resolvers: {
        Query: {
            vwFixtures:  (parent, fieldArgs, cache, info) => {
                const result = vxFixtureResolver(parent, fieldArgs, cache, info);
                console.log(result, JSON.stringify(info, null, 2))
                return result;
            }
        }
    }
});

const client = createClient({
    url: import.meta.env.VITE_APP_HASURA ? import.meta.env.VITE_APP_HASURA : "",
    exchanges: [
        cache,
        devtoolsExchange,
        // debugExchange,
        dedupExchange,
        fetchExchange,
    ]
});
```

On a side note: When retrieving paginated data from the cache, what is the correct behavior in terms of the amount data to be returned? Currently, everything is returned at once but prior to it being cached, the data is returned in sets of 20, which is what I've set it to.
